### ISSUE TITLE
Account for multiple meeting minutes

### ISSUE THREAD

**Issue by antidipyramid** [Metro-Records/scrapers-lametro#16]
Per @camachoo:

> Can we figure out why the March 28, 2024, meeting minutes are not showing up properly? (BTW, I think I already know why this is happening.) They were entered and approved on the April 25, 2024, meeting but as part of a double minutes item. We need to have some logic put in place that can account for more than one minutes attachment on the minutes item type. The Regular Board Meeting minutes follow the same title format “Regular Board Meeting MINUTES - Month ##, ####” and the Special Board Budget Workshop uses this title format “Special Board Budget Workshop MINUTES - Month ##, ####”.  For the budget workshop, can we look for “Board Budget Workshop MINUTES” in the logic?
>
> Suggestions are welcome.  Before you ask, separate items for each minutes item every time is not an option.  So, we will have to workshop a way to accommodate the minutes items as attachments on a single minutes board reports and separate minutes board reports of the following month.

**Comment by antidipyramid** [comment]
@xmedr will look into analyzing file structure as a failsafe for distinguishing minutes files.

**Comment by xmedr** [comment]
### Some observations about minutes in legistar:
- There a few different structures for minutes files, though they always have the word "MINUTES" in all caps on the first page
- All files with a `MatterTypeName` of "Minutes" have the word minutes in the name and seem to contain minutes
- Some matters with minutes in the name can also have a matter type of `Informational Report` instead of `Minutes`. Looking at those files shows that they are still minutes files.
- One matter with a type of "Motion / Motion Response" contained the word minutes in the name but did not contain a minutes file. However, the full name in context did not identify this as a minutes file to begin with, so this type can likely be ignored
  - The full name was: "APPROVE Motion by Fasana that Metro report back to the Operations Committee by January 2020 with a table for each rail line showing the number of occurrences gates have been down at each protected intersection for longer than 3 minutes, 5 minutes, and 10 minutes."

### Takeaway
This might mean that we can look at `matters with a mattertype of minutes` OR `matters with minutes in the name and a mattertype of informational report`, and look at the attachments within to see if they contain the word MINUTES in all caps as a final check. If all works out, then save the attachment as a minutes file. 

```
# A list of all the matters that match the above conditions
https://webapi.legistar.com/v1/metro/matters/?$filter=(MatterTypeName eq 'Minutes') or (substringof('Minutes', MatterTitle) and MatterTypeName eq 'Informational Report')
```

What do you think about all this @antidipyramid @hancush? The only caveat I can think of is that the process would end up taking longer than it currently does since we'd be looking at each attachment to double check that it's the right kind of file.

---

Examples of different minutes file structures:
- https://metro.legistar1.com/metro/attachments/135464c6-8ad1-4df8-9f89-cfa60b569f02.pdf
- https://metro.legistar1.com/metro/attachments/548a0133-229a-4aa9-8ae7-074288c89300.pdf
- https://metro.legistar1.com/metro/attachments/d21e1c47-47dc-4f52-9a27-5b377bd4798e.pdf
- https://metro.legistar1.com/metro/attachments/9527596e-5640-4a47-beb1-1ae7331fdcfc.pdf
- https://metro.legistar1.com/metro/attachments/c1c46dd4-b06a-413e-9e00-04342f1c5b7e.pdf

**Comment by hancush** [comment]
@xmedr Thanks for the spelunking!

Re: processing time, we can skip the more involved check if there is a minutes file explicitly attached or denoted as not available. See:

https://github.com/Metro-Records/scrapers-lametro/blob/a660f649f09984470182807e61e9e7147cadea2d/lametro/events.py#L391-L417

We can also skip peeking at the document if there is only one attachment resembling minutes to a minutes board report, i.e., it's unambiguous which attachment contains minutes.

https://github.com/Metro-Records/scrapers-lametro/blob/a660f649f09984470182807e61e9e7147cadea2d/lametro/events.py#L531-L580

I think strategy you propose would make a great update to the `find_approved_minutes` method. Note that the current implementation assumes that there will only ever be one board report with minutes in its attachments. Per this issue, there might be more than one, so we'll need to accommodate that. Likewise, it assumes that an event will only ever have one minutes file. We'll need to revise that, too.

Want to take a stab at revising `find_approved_minutes`? (Also happy to pair, if you prefer!)

**Comment by xmedr** [comment]
This is great, tysm! I can definitely get started on those revisions

**Comment by hancush** [comment]
First draft of this is done, but it hasn't been deployed to production yet. 

@xmedr Can you summarize how this revised approach has been doing since we deployed it to staging last week?

**Comment by xmedr** [comment]
Absolutely!

Upon deploying, we found that there was a matter that is now being scraped that does not contain attachments and is raising an error because of it. The matter id is `7103` and the link it's accessing is: https://webapi.legistar.com/v1/metro/matters/7103/attachments

Part of the change was that we modified the matter search from those with the word Minutes in the title, to those where the matter type is minutes OR the matter type is informational report and the the word Minutes is in the title.

Upon checking out [the exact search that the scraper is now using in the browser](https://webapi.legistar.com/v1/metro/matters/?$filter=MatterBodyId%20eq%20138%20and%20substringof(%27September%2024,%202020%27,%20MatterTitle)%20and%20((MatterTypeName%20eq%20%27Minutes%27)%20or%20(substringof(%27Minutes%27,%20MatterTitle)%20and%20MatterTypeName%20eq%20%27Informational%20Report%27))), we noticed that the problem matter isn't listed but a different one timestamped very closely to the problematic one is present. Then, [looking at that matter in the scraper](https://gist.github.com/xmedr/d228bb089ad87bf56a117da2ce63afbc), we saw that it's grabbing a draft matter that truly does not have attachments.

Edit: Did not mean to re-close the issue!

**Comment by hancush** [comment]
@xmedr Thanks! Is this production ready, then, or does a change need to be made to accommodate the edge case you describe?

**Comment by xmedr** [comment]
@hancush I'll be working on some changes this week! And this will stay on staging until we've got a solution ready for those kinds of matters.

**Comment by hancush** [comment]
Gotcha! In that case, can you take the change off `main`, so that it isn't in the way if we need to make other (potentially urgent) fixes?

To do that:

1. Branch off `main`.
2. Revert your merge commit: `git revert -m 1 ${SHA OF THE MERGE COMMIT}`
3. Push the revert commit to your branch and submit a PR.
4. Once your PR is merged, branch off `main` again and revert your revert: `git revert -m 1 ${SHA OF THE REVERT COMMIT}`
5. Submit a PR with your change reapplied (the result of reverting the revert).

It's a convoluted process, to be sure! 🫠

**Comment by xmedr** [comment]
On this most recent version, we're programmatically skipping restricted matters. The first time this ran on staging, we encountered an error on the event scrape when it encountered the following link: https://webapi.legistar.com/v1/metro/matters/9112/attachments

The error itself read: `pdfminer.pdfparser.PDFSyntaxError : No /Root object! - Is this really a PDF?` which seems to be raised when the pdf is configured incorrectly in some way as per [this issue in the pdfminer's repo](https://github.com/pdfminer/pdfminer.six/issues/476). Opening the attachments in that link worked fine, and running a local event scrape with a try block didn't raise any errors. Since then, the event scrape has completed twice without error when coming across that same link above. We can see some [events on councilmatic staging that now have multiple minutes](https://la-metro-councilmatic-staging.herokuapp.com/event/regular-board-meeting-dee60ad2a49b/)!

I imagine I could either continue to monitor the staging event scrape to see if this error happens again, or push up a try block to catch that exception and report it to sentry. What do you think @hancush?

(Side note: the full scrape itself on staging has been failing on the bills scrape the last few runs with and without this change merged in, but it seems to be because of a general timeout )

**Comment by hancush** [comment]
Love the Sentry idea, @xmedr!

**Comment by xmedr** [comment]
@hancush Sweet, can do! And would it be better to continue the scrape after reporting or still let it error out?

**Comment by hancush** [comment]
@xmedr Continue the scrape.

**Comment by xmedr** [comment]
Solution deployed to both the scrapers and councilmatic, and viewable on the meeting cited in the original comment: https://boardagendas.metro.net/event/regular-board-meeting-dee60ad2a49b/

![image](https://github.com/user-attachments/assets/0cb36453-44b3-4d03-9130-53a085047c17)
